docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+(e.g. by kube-proxy).
+For example, if we were using AWS subnets to identify ELBs (which we're not), the external load-balancer name
+that the k8s-end-user cares about would be something like `myelbname.amazonaws.com`, the internal ip-ranges would be
+something like `10.244.250.192/30`.  In the case of GCE these are the same, and it looks like we don't need to track
+an internal name at the moment, but we do not want to bake this assumption into the API.  So publicIPs is _external_.
+
+publicIPs in the ServiceSpec will continue to work as it does today. The user can specify a publicIP to use an existing
+load-balancer/VIP (in practice this only works for OpenStack today).  This will be populated once the load-balancer
+is created.  This can be deprecated in future (once the OpenStack plugin does dynamic VIP creation).
+
+Each cloud-provider can keep track of the mapping from service to load-balancer:
+* GCE uses the name
+* AWS can use name and/or tags
+* OpenStack uses the name
+* Rackspace can use the name and/or metadata
+
+Because the load-balancer IPs can be recreated, we are able to expose the load-balancer information through the status.
+The ServiceStatus will be extended to include publicIPs.  This will be the user-facing IP addresses or hostnames.
+
+If we require additional state in future for kube-proxy routing (for example internal load-balancer IPs),
+that will be added to a new Status field.
+
+By putting this information on Status, an external load-balancer provisioner can use the same mechanism for securely
+updating Status as kube-proxy/kubelet.
+
+
+
+## Creation flow
+
+### visibility: loadbalancer
+
+1. User `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Spec{ visibility: loadbalancer, ports: [{port: 80}] } }`
+1. API server assigns a hostPort (if the load balancer needs one), accepts and persists
+1. LB Controller wakes up, allocates a load-balancer, and `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Status{ publicIps: [ "1.2.3.4" ] }}`
+1. Service REST sets `Service.Status.publicIps = [ "1.2.3.4" ]`
+1. kube-proxy wakes up and sets iptables to receive on 1.2.3.4
+
+### visibility: public
+
+1. User `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Spec{ visibility: public, ports: [{port: 80, hostPort: 10080}] } }`
+1. API server verifies that port 10080 can be assigned, accepts and persists
+1. kube-proxy wakes up and sets iptables to receive on hostPort 10080
+
+
+In order to verify that the port can be assigned, the API server must check that the port is within a permitted range,
+and that it is not already assigned to another service.  This is similar to the allocation of internal cluster-IPs.
+The error codes & messages (in the case of port conflicts, out-of-range ports or port-exhaustion) will be analogous
+to the equivalent IP address errors.
+
+Advanced clouds (e.g. GCE) may support load-balancing without requiring a public port assignment.
```

Comment:
Yeah, I think that by having loadbalancer => public this is easier.  I don't love the fact that you can bypass the load-balancer, but I think this works.

I was thinking that load-balancer and public would be "joint-first", i.e. you would choose public or loadbalancer, but loadbalancer => public is just easier.